File: ruby/raindrops/raindrops.rb
module Raindrops
  class << self
    def convert(n)
      acc = ""
      acc = maybe_add_pling(acc, n)
      acc = maybe_add_plang(acc, n)
      acc = maybe_add_plong(acc, n)
      acc.size == 0 ? n.to_s : acc
    end

    {pling: 3, plang: 5, plong: 7}.each do |(sound, div)|
      class_eval <<-META
      def maybe_add_#{sound}(acc, n)
        n % #{div} == 0 ? acc << "#{sound.capitalize}" : acc
      end
      META
    end
  end
end

File: erlang/raindrops/src/raindrops.erl
-module(raindrops).

-export([convert/1, test_version/0]).

convert(Number) ->
  String = "",
  String2 = maybe_add_pling(Number, String),
  String3 = maybe_add_plang(Number, String2),
  String4 = maybe_add_plong(Number, String3),
  maybe_add_integer(Number, String4).

maybe_add_pling(Number, Acc) when Number rem 3 == 0 ->
  Acc ++ "Pling";
maybe_add_pling(_, Acc) ->
  Acc.

maybe_add_plang(Number, Acc) when Number rem 5 == 0 ->
  Acc ++ "Plang";
maybe_add_plang(_, Acc) ->
  Acc.

maybe_add_plong(Number, Acc) when Number rem 7 == 0 ->
  Acc ++ "Plong";
maybe_add_plong(_, Acc) ->
  Acc.

maybe_add_integer(Number, Acc) when length(Acc) == 0 ->
  integer_to_list(Number);
maybe_add_integer(_, Acc) ->
  Acc.

test_version() -> 1.
